### ISSUE TITLE
Exposing `publishAllOutstanding` as a public API in `PublisherInterface`

### ISSUE THREAD

**Issue by elefeint** [googleapis/java-pubsub#949]
In the last release of Spring Cloud GCP, we switched from returning `Publisher` to returning `PublisherInterface` in the convenience "Template" layer. This led to an accidental regression (GoogleCloudPlatform/spring-cloud-gcp#790).

Before we proceed with workarounds in Spring Cloud GCP, an API surface question for the client library --  does it make sense to add `publishAllOutstanding()` to `PublisherInterface`? Or is it not meant to be part of the public API?

**Comment by meredithslota** [comment]
@kamalaboulhosn @hannahrogers-google @hongalex @feywind for discussion.

**Comment by meredithslota** [comment]
Also @anguillanneuf for discussion.

**Comment by anguillanneuf** [comment]
I have not used `publishAllOutstanding()` anywhere in samples. Deferring this to the other folks for comment.

**Comment by kamalaboulhosn** [comment]
The PublisherInterface is [used by the Pub/Sub Lite library](https://github.com/googleapis/java-pubsublite/blob/6a9e041099af72dab7ffb01a29ce658413fbd369/google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/Publisher.java#L24). It does not implement the `publishAllOutstanding()` method, though. Therefore, adding the method to the interface would break the Pub/Sub Lite client library unless we added support for this method. I'm not sure how well that fits into Lite. Perhaps @dpcollins-google has some thoughts.

What was the reason for switching from `Publisher` to `PublisherInterface`?

**Comment by meltsufin** [comment]
> What was the reason for switching from `Publisher` to `PublisherInterface`?

It made it possible to add trace instrumentation here: https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/548. We basically needed a `TracingPublisher` that can be used in place of `Publisher`.

Note that it's impossible to subclass `Publisher` because it only has one private constructor.

**Comment by elefeint** [comment]
Happy New Year! I am going to try switching Spring Cloud GCP to use `Publisher.setTransform()` instead of a custom implementation of `PublisherInterface` to add tracing information, since the need to interact with the full set of `Publisher` functionality came up once again in GoogleCloudPlatform/spring-cloud-gcp#864.

@kamalaboulhosn Could you confirm/clarify a couple of my assumptions --
1) `PublisherInterface` is not meant to be the way to interact with `Publisher` functionality, and so is not expected to expose all public API methods of `Publisher`.
2) `Publisher.setTransform()` is okay to use. It has `@BetaApi` on it, so I want to double check. The functionality was added for OpenCensus support, so it fits our usecase perfectly.

**Comment by hannahrogers-google** [comment]
Hi @elefeint,

The only reason we added the PublisherInterface was because we created a Pub/Sub Lite service, as Kamal mentioned. The interface only exposes methods that are implemented by both the Pub/Sub Lite publisher client and the Cloud Pub/Sub publisher client to allow for seamless transition between Cloud Pub/Sub and Pub/Sub Lite. The other methods exposed by the Cloud Pub/Sub publisher client (like publishAllOutstanding) are not implemented by the Pub/Sub Lite client, which is why these methods are not included in the PublisherInterface.

Publisher.setTransform is okay to use, even though it has the BetaApi tag.
